src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/DependenciesGraphProvider.cs

```diff
@@ -0,0 +1,653 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.GraphModel;
+using Microsoft.VisualStudio.GraphModel.CodeSchema;
+using Microsoft.VisualStudio.GraphModel.Schemas;
+using Microsoft.VisualStudio.Imaging.Interop;
+using Microsoft.VisualStudio.PlatformUI;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Threading;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Tree.Dependencies
+{
+    /// <summary>
+    /// Provides actual dependencies nodes under Dependencies\[DependencyType]\[TopLevel]\[....] sub nodes. 
+    /// </summary>
+    [GraphProvider(Name = "Microsoft.VisualStudio.ProjectSystem.VS.Tree.DependenciesNodeGraphProvider")]
+    // TODO what todo with ProjectKind here? Provide several implementations or request change to provide multiple project types?
+    //      ProjectKind = ProjectSystemPackage.ProjectTypeGuidStringWithParentheses)]
+    // without specifying ProjectType our GraphProvider will be called for all project types
+    // which might cause some extra perf overhead.
+    internal class DependenciesGraphProvider : OnceInitializedOnceDisposedAsync, IGraphProvider
+    {
+        private IDependenciesGraphProjectContextProvider ProjectContextProvider { get; set; }
+        private SVsServiceProvider ServiceProvider { get; set; }
+
+        private bool _IsInitialized;
+
+        private object _initializationLock = new object();
+
+        [ImportingConstructor]
+        public DependenciesGraphProvider(IDependenciesGraphProjectContextProvider projectContextProvider,
+                                             SVsServiceProvider serviceProvider)
+            : base(new JoinableTaskContextNode(ThreadHelper.JoinableTaskContext))
+        {
+            ProjectContextProvider = projectContextProvider;
+            ServiceProvider = serviceProvider;
+        }
+
+        /// <summary>
+        /// Remembers expanded graph nodes.
+        /// TODO: Check how to clean them when project is unloaded. (What graph is doing in general when project is killed?)
+        /// </summary>
+        private WeakCollection<IGraphContext> ExpandedGraphContexts { get; set; }
+        private HashSet<string> RegisteredSubTreeProviders { get; set; }
+
+        private object _RegisteredSubTreeProvidersLock = new object();
+
+        private object _ExpandedGraphContextsLock = new object();
```

Comment:
readonly,

We put fields before constructors, and properties after constructors.